pkg/controller/common.go

```diff
@@ -597,335 +600,805 @@ func (c *SharedState) isVolumeVisited(azVolumeName string) bool {
+							klog.Errorf("failed to retrieve pod list: %v", err)
+							continue
+						}
+						pods = append(pods, podList.Items...)
+					}
+				} else {
+					podList := &v1.PodList{}
+					if err = p.azr.getClient().List(ctx, podList, &client.ListOptions{LabelSelector: podAffinities}); err != nil {
+						klog.Errorf("failed to retrieve pod list: %v", err)
+						continue
+					}
+					pods = podList.Items
+				}
+
+				nodesWithSelectedPods := set{}
+				for _, pod := range pods {
+					klog.Infof("Pod (%s) has matching label for pod affinity (%v)", getQualifiedName(pod.Namespace, pod.Name), podAffinities)
+					nodesWithSelectedPods.add(pod.Spec.NodeName)
+				}
+
+				// add nodes, to which replica attachments of matching pods' volumes are attached, to replicaNodes
+				if volumes, err := p.azr.getSharedState().getVolumesForPodObjs(pods...); err == nil {
+					for _, volume := range volumes {
+						attachments, err := getAzVolumeAttachmentsForVolume(ctx, p.azr.getClient(), volume, replicaOnly)
+						if err != nil {
+							continue
+						}
 
-			updatedTolerations := []v1.Toleration{}
-			for _, podToleration := range podTolerations {
-				if existingToleration, ok := podTolerationMap[podToleration.Key]; ok {
-					if podToleration.MatchToleration(existingToleration) {
-						updatedTolerations = append(updatedTolerations, podToleration)
+						nodeChecker := set{}
+						for _, attachment := range attachments {
+							if i == 0 {
+								klog.V(5).Infof("Adding node (%s) to the replica node set: replica mounts of volume (%s) found on node", attachment.Spec.NodeName, volume)
+								replicaNodes.add(attachment.Spec.NodeName)
+							} else {
+								nodeChecker.add(attachment.Spec.NodeName)
+							}
+						}
+						if i > 0 {
+							// take an intersection of the current pod list's replica nodes with those of preceding pod lists.
+							for node := range replicaNodes {
+								if !nodeChecker.has(node) {
+									klog.V(5).Infof("Removing node (%s) from the replica node set: replica mounts of volume (%s) cannot be found on node", node, volume)
+									replicaNodes.remove(node)
+								}
+							}
+						}
+					}
+				}
+
+				qualifiedNodes := set{}
+				for node := range nodesWithSelectedPods {
+					var nodeInstance v1.Node
+					if err = p.azr.getClient().Get(ctx, types.NamespacedName{Name: node.(string)}, &nodeInstance); err != nil {
+						klog.Errorf("failed to get node (%s)", node.(string))
+						continue
+					}
+					if value, exists := nodeInstance.Labels[podAffinity.TopologyKey]; exists {
+						// get list of node candidates
+						nodeRequirement, err := azureutils.CreateLabelRequirements(podAffinity.TopologyKey, selection.Equals, value)
+						if err != nil {
+							continue
+						}
+						nodeSelector := labels.NewSelector().Add(*nodeRequirement)
+
+						nodeList := &v1.NodeList{}
+						if err = p.azr.getClient().List(ctx, nodeList, &client.ListOptions{LabelSelector: nodeSelector}); err != nil {
```

Comment:
Instead of calling the List function for every node in nodesWithSelectedPods, can't we add the label requirements for every node in nodesWithSelectedPods, to the nodeSelector within the loop, and then call List function outside of the nodesWithSelectedPods  loop?

Reply:
The problem is that if node 1 for pod A and node 2 for pod B have different value for the selected label, list function will not be able to OR these values for fetching qualifying nodes.

Reply:
I guess we can use `selection.In` to handle this case. That way, we can avoid making multiple list calls.

Reply:
Yeah I was just writing that :D

Reply:
Also, like I mentioned in other comments, can't we just filter out from the candidate nodes using nodeSelector instead of making a List call at all?

Reply:
Yeah, thinking twice, that should be work 👍